### ISSUE TITLE
How to train with Online Hard Example Mining?

### ISSUE THREAD

**Issue by John1231983** [scaelles/DEXTR-PyTorch#2]
Thanks for your great work. This is not an issue/bug, I just want to make clear what you did with the hard sample. I think it will be useful for other researchers as well. In the paper, you mentioned

> We **split the training dataset** into two equal splits. ..We first select the objects that lead to poor performance (IoU<0.8) when applying the network trained on the first split... Using the network trained further on the hard examples...
Since the increased performance is partially due to the increased amount of training data (first split + hard examples of the second split), we need to **disentangle the two sources** of performance gain. 

Assume we have `1000` images in the first split and `50` hard examples, so we will merge them to generate `1050` images and train with initial checkpoint from the trained model. So, if the learning rate of the first split is `[0.01 0.001 0.0001]` to obtained the trained model for the first split (without hard sample) and early stop at `0.001`. What should a learning rate in the second phase (first split+hard sample)? Does it start again from  `[0.01 0.001 0.0001]` or the last learning rate `0.001`?

**Comment by kmaninis** [comment]
Hi, thanks for your comments.
In the paper, we did not change the learning rate, but rather kept it fixed, so there was no such issue.

However, what I would do in the case described above is start again from the beginning, since the task (5 points instead of 4) is slightly different.

**Comment by John1231983** [comment]
Thanks for your reply. One more thing I want to mention here is that your scheme likes bootstrapping with the hard example, rather than OHEM. Because you have stopped training and add hard sample to the first split by hand, then retrain them. Am I right?

**Comment by kmaninis** [comment]
Ah, sorry for the confusion.
In the second phase, we train **only** on the hard examples, and that's the relation to OHEM.
The process is:
1.  Score the image with 4 extreme points as input.
2.  Compute the most erroneous location.
3.  Append the most erroneous point to the extreme points and score with 5 points.
4.  Compare with the ground truth and backpropagate.
This is to simulate human behaviour for interactive segmentation, when the initial result is failing.

I hope this helps!

**Comment by Valentine233** [comment]
Hi, I have a question about "interactive object segmentation evaluation".

For my understanding, you use all test set (second split) to train "5points" while using only hard examples to train "5points + OHEM". The final test set for calculating IoU is also the second split.

Since "5points" model has seen all samples in test set, why would it get a worse performance?

**Comment by scoobym** [comment]
So how do you compute  the most erroneous location and get the the most erroneous point? I don't get it. Does the code you provide here include this?

**Comment by kmaninis** [comment]
It is all simulated, we know the ground-truth mask, and compute the most erroneous location (== wrong prediction furthest from the true boundary).
